src/openboardview/FileFormats/BVR3File.cpp

```diff
@@ -0,0 +1,207 @@
+#include "BVR3File.h"
+
+#include "utils.h"
+#include <cmath>
+#include <cctype>
+#include <clocale>
+#include <cstdint>
+#include <cstring>
+#include <list>
+
+int manhattan_distance(const BRDPoint &p1, const BRDPoint &p2) {
+	return abs(p1.x - p2.x) + abs(p1.y - p2.y);
+}
+
+bool BVR3File::verifyFormat(std::vector<char> &buf) {
+	return find_str_in_buf("BVRAW_FORMAT_3", buf);
+}
+
+BVR3File::BVR3File(std::vector<char> &buf) {
+	auto buffer_size = buf.size();
+
+	char *saved_locale = setlocale(LC_NUMERIC, "C"); // Use '.' as delimiter for strtod
+
+	ENSURE_OR_FAIL(buffer_size > 4, error_msg, return);
+	size_t file_buf_size = 3 * (1 + buffer_size);
+	file_buf             = (char *)calloc(1, file_buf_size);
+	ENSURE_OR_FAIL(file_buf != nullptr, error_msg, return);
+
+	std::copy(buf.begin(), buf.end(), file_buf);
+	file_buf[buffer_size] = 0;
+	// This is for fixing degenerate utf8
+	char *arena     = &file_buf[buffer_size + 1];
+	char *arena_end = file_buf + file_buf_size - 1;
+	*arena_end      = 0;
+
+	BRDPart blank_part;
+	BRDPin blank_pin;
+	BRDPart part;
+	BRDPin pin;
+	std::list<std::pair<BRDPoint, BRDPoint>> outline_segments;
+
+	std::vector<char *> lines;
+	stringfile(file_buf, lines);
+
+	for (char *line : lines) {
+		while (isspace((uint8_t)*line)) line++;
+		if (!line[0]) continue;
+
+		char *p = line;
+		char *s;
+
+		if (!strncmp(line, "PART_NAME ", 10)) {
+			p += 10;
+			part.name = READ_STR();
+		} else if (!strncmp(line, "PART_SIDE ", 10)) {
+			p += 10;
+			char *side = READ_STR();
+			if (!strcmp(side, "T"))
+				part.mounting_side = BRDPartMountingSide::Top;
+			else
+				part.mounting_side = BRDPartMountingSide::Bottom;
+		} else if (!strncmp(line, "PART_ORIGIN ", 12)) {
+			p += 12;
+			double origin_x = READ_DOUBLE();
+			part.p1.x = trunc(origin_x);
+			double origin_y = READ_DOUBLE();
+			part.p1.y = trunc(origin_y);
+		} else if (!strncmp(line, "PART_MOUNT ", 11)) {
+			p += 11;
+			char *mount = READ_STR();
+			if (!strcmp(mount, "SMD"))
+				part.part_type = BRDPartType::SMD;
+			else
+				part.part_type = BRDPartType::ThroughHole;
+		} else if (!strncmp(line, "PIN_ID ", 7)) {
+			p += 7;
+			pin.snum = READ_STR();
+			pin.part = parts.size() + 1; // pin is for current part, which will not yet have been added to parts vector
+		} else if (!strncmp(line, "PIN_NUMBER ", 11)) {
+			// Ignored, not sure if relevant for BRDPin
+		} else if (!strncmp(line, "PIN_NAME ", 9)) {
+			p += 9;
+			pin.name = READ_STR();
+		} else if (!strncmp(line, "PIN_SIDE ", 9)) {
+			// Ignored, not sure if relevant for BRDPin
```

Comment:
I don't think `BRDPin::side` is used right now. Could be used in case a part has pins on different sides (or both).

Reply:
I've looked back at this and I think the reason I didn't populate that was because I didn't know how T and B values from the file translate to BRDPin::side, which is a uint. If you can advise, I'll modify to suit.

Reply:
`BRDPin::side` is not really used for anything right now, only assigned in `BRD2File.cpp` with 1 → Top, 2 → Bottom.
In `BRDBoard.cpp` for `BRDNail::side`, 1 → Top, everything else → Bottom.
In `Board.h`, `EBoardSide` for `BoardElement`, 0 → Top, 1 → Bottom, 2 → Both.
It's not consistent in any case. There should be only one `enum` (`enum class` may I say) for everything.
Not useful for now if there are no files that contain a part with pins on different sides.

Reply:
Thanks for that info. I've modified the code to treat things the same way as BRD2File, but at some stage it will want replacing with enum values that are consistent across the codebase.